porch/apiserver/pkg/registry/porch/table.go

```diff
@@ -0,0 +1,150 @@
+	api "github.com/GoogleContainerTools/kpt/porch/api/porch/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+)
+
+type tableConvertor struct {
+	resource schema.GroupResource
+	// cells creates a single row of cells of the table from a runtime.Object
+	cells func(obj runtime.Object) []interface{}
+	// columns stores column definitions for the table convertor
+	columns []metav1.TableColumnDefinition
+}
+
+var _ rest.TableConvertor = tableConvertor{}
+
+// ConvertToTable implements rest.TableConvertor
+func (c tableConvertor) ConvertToTable(ctx context.Context, object runtime.Object, tableOptions runtime.Object) (*metav1.Table, error) {
+	var table metav1.Table
+
+	fn := func(obj runtime.Object) error {
+		cells := c.cells(obj)
+		if cells == nil || len(cells) == 0 {
+			return notAcceptableError(ctx, c.resource)
+		}
+		table.Rows = append(table.Rows, metav1.TableRow{
+			Cells:  cells,
+			Object: runtime.RawExtension{Object: obj},
+		})
+		return nil
+	}
+
+	// Create table rows
+	switch {
+	case meta.IsListType(object):
+		if err := meta.EachListItem(object, fn); err != nil {
+			return nil, err
+		}
+	default:
+		if err := fn(object); err != nil {
+			return nil, err
+		}
+	}
+
+	// Populate table metadata
+	if l, err := meta.ListAccessor(object); err == nil {
+		table.ResourceVersion = l.GetResourceVersion()
+		table.SelfLink = l.GetSelfLink()
+		table.Continue = l.GetContinue()
+		table.RemainingItemCount = l.GetRemainingItemCount()
+	} else if c, err := meta.CommonAccessor(object); err == nil {
+		table.ResourceVersion = c.GetResourceVersion()
+		table.SelfLink = c.GetSelfLink()
+	}
+	if opt, ok := tableOptions.(*metav1.TableOptions); !ok || !opt.NoHeaders {
+		table.ColumnDefinitions = c.columns
+	}
+
+	return &table, nil
+}
+
+var (
+	packageRevisionTableConvertor = tableConvertor{
+		resource: porch.Resource("packagerevisions"),
+		cells: func(obj runtime.Object) []interface{} {
+			pr, ok := obj.(*api.PackageRevision)
+			if !ok {
+				return nil
+			}
+			return []interface{}{
+				pr.Name,
+				pr.Spec.PackageName,
+				pr.Spec.Revision,
+				isLatest(pr),
+				pr.Spec.Lifecycle,
+				pr.Spec.RepositoryName,
+			}
+		},
+		columns: []metav1.TableColumnDefinition{
+			{Name: "Name", Type: "string"},
+			{Name: "Package", Type: "string"},
+			{Name: "Revision", Type: "string"},
+			{Name: "Latest", Type: "boolean"},
+			{Name: "Lifecycle", Type: "string"},
+			{Name: "Repository", Type: "string"},
+		},
+	}
+)
+
+func isLatest(pr *api.PackageRevision) bool {
+	labels := pr.Labels
+	if labels == nil {
+		return false
+	}
+	return labels[api.LatestPackageRevisionKey] == api.LatestPackageRevisionValue
+}
+
+func notAcceptableError(ctx context.Context, resource schema.GroupResource) error {
```

Comment:
Nit (for future PR): move these into their own file (errors.go)?

Reply:
did it now. thanks!